test/e2e/basic_test.go

```diff
@@ -49,165 +51,188 @@ func TestGetStatsSummary(t *testing.T) {
+}
+
+// TestPodLifecycleNonGracefulDelete creates one podsand verifies that the provider has created them
+// and put them in the running lifecycle. It then does a force delete on the pod, and verifies the provider
+// has deleted it.
+func TestPodLifecycleForceDelete(t *testing.T) {
+	podSpec := f.CreateDummyPodObjectWithPrefix(t.Name(), "nginx-", "foo")
+	// Create a pod with prefix having a single container.
+	pod, err := f.CreatePod(podSpec)
 	if err != nil {
 		t.Fatal(err)
 	}
+	// Delete the pod after the test finishes.
+	defer func() {
+		if err := f.DeletePodImmediately(pod.Namespace, pod.Name); err != nil && !apierrors.IsNotFound(err) {
+			t.Error(err)
+		}
+	}()
+	t.Logf("Created pod: %s", pod.Name)
 
-	// Make sure the "nginx-1-Y" pod DOES NOT exist in the slice of PodStats anymore.
-	if _, err := findPodInPodStats(stats, pod1); err == nil {
-		t.Fatalf("expected to NOT find pod \"%s/%s\" in the slice of pod stats", pod1.Namespace, pod1.Name)
+	// Wait for the "nginx-" pod to be reported as running and ready.
+	if _, err := f.WaitUntilPodReady(pod.Namespace, pod.Name); err != nil {
+		t.Fatal(err)
 	}
+	t.Logf("Pod %s ready", pod.Name)
+
+	// Grab the pods from the provider.
+	pods, err := f.GetRunningPods()
+	assert.NilError(t, err)
+
+	// Check if the pod exists in the slice of Pods.
+	assert.NilError(t, findPodInPods(pods, pod))
 
-	// Wait for the "nginx-0-X" pod to be deleted in a separate goroutine.
+	// Wait for the pod to be deleted in a separate goroutine.
 	// This ensures that we don't possibly miss the MODIFIED/DELETED events due to establishing the watch too late in the process.
-	pod0Ch := make(chan error)
+	// It also makes sure that in light of soft deletes, we properly handle non-graceful pod deletion
+	podCh := make(chan error)
+	var podLast *v1.Pod
 	go func() {
-		// Wait for the "nginx-0-X" pod to be reported as having been deleted.
-		if err := f.WaitUntilPodDeleted(pod0.Namespace, pod0.Name); err != nil {
+		// Close the podCh channel, signaling we've observed deletion of the pod.
+		defer close(podCh)
+
+		var err error
+		// Wait for the pod to be reported as having been deleted.
+		podLast, err = f.WaitUntilPodDeleted(pod.Namespace, pod.Name)
+		if err != nil {
 			// Propagate the error to the outside so we can fail the test.
-			pod0Ch <- err
-		} else {
-			// Close the pod0Ch channel, signaling we've observed deletion of the pod.
-			close(pod0Ch)
+			podCh <- err
 		}
 	}()
 
-	// Forcibly delete the "nginx-0" pod.
-	if err := f.DeletePodImmediately(pod0.Namespace, pod0.Name); err != nil {
+	time.Sleep(deleteGracePeriodForProvider)
+	// Forcibly delete the pod.
+	if err := f.DeletePodImmediately(pod.Namespace, pod.Name); err != nil {
+		t.Logf("Last saw pod in state: %+v", podLast)
 		t.Fatal(err)
 	}
+	t.Log("Force deleted pod: ", pod.Name)
+
 	// Wait for the delete event to be ACKed.
-	if err := <-pod0Ch; err != nil {
+	if err := <-podCh; err != nil {
```

Comment:
Select with timeout?
I haven't looked at `WaitUntilPodDeleted` but it would be nice if that took a cancellable context.